libs/motor_control/esc/esc.c

```diff
@@ -0,0 +1,164 @@
+/*
+ * This file is part of the efc project <https://github.com/eurus-project/efc/>.
+ * Copyright (c) 2024-2025 The efc developers.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, version 3.
+ *
+ * This program is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "esc.h"
+#include <stdint.h>
+#include <stdlib.h>
+#include <zephyr/devicetree.h>
+#include <zephyr/devicetree/pinctrl.h>
+#include <zephyr/drivers/pwm.h>
+#include <zephyr/kernel.h>
+
+ESC_Error_Type ESC_Init(ESC_Inst_Type *esc_out, const struct device *pwm_dev,
+                        uint32_t pwm_channel, ESC_Protocol_Type protocol) {
+    static uint8_t esc_instance_cnt = 0;
+
+    if (!device_is_ready(pwm_dev))
+        return ESC_DEVICE_PWM_NOT_READY;
+
+    switch (protocol) {
+    case ESC_PWM:
+        esc_out->period = ESC_T_PWM_US;
+        esc_out->dc_min = ESC_DC_MIN_PWM_US;
+        esc_out->dc_max = ESC_DC_MAX_PWM_US;
+        esc_out->protocol = ESC_PWM;
+        break;
+
+    case ESC_ONESHOT_125:
+        esc_out->period = ESC_T_ONESHOT_125;
+        esc_out->dc_min = ESC_DC_MIN_ONESHOT_125_US;
+        esc_out->dc_max = ESC_DC_MAX_ONESHOT_125_US;
+        esc_out->protocol = ESC_ONESHOT_125;
+        break;
+
+    case ESC_ONESHOT_42:
+        esc_out->period = ESC_T_ONESHOT_42;
+        esc_out->dc_min = ESC_DC_MIN_ONESHOT_42_US;
+        esc_out->dc_max = ESC_DC_MAX_ONESHOT_42_US;
+        esc_out->protocol = ESC_ONESHOT_42;
+        break;
+
+    case ESC_MULTISHOT:
+        esc_out->period = ESC_T_MULTISHOT;
+        esc_out->dc_min = ESC_DC_MIN_MULTISHOT_US;
+        esc_out->dc_max = ESC_DC_MAX_MULTISHOT_US;
+        esc_out->protocol = ESC_MULTISHOT;
+        break;
+
+    default:
+        esc_out->flag = false;
+        return ESC_NOT_INITIALIZED;
+    }
+    esc_out->pwm_dev = pwm_dev;
+    esc_out->pwm_channel = pwm_channel;
+    esc_out->instance_count = esc_instance_cnt;
+    esc_out->flag = true;
+    esc_instance_cnt++;
+
+    return ESC_SUCCESS;
+}
+
+ESC_Error_Type ESC_SetSpeed(ESC_Inst_Type *esc, float speed) {
+    int ret;
+    uint32_t dc;
+
+    if (!esc->flag) {
+        return ESC_NOT_INITIALIZED;
+    }
+
+    if (speed > 1.0f || speed < 0.0f)
+        return ESC_INVALID_SPEED_SET;
+
+    dc = (uint32_t)(ESC_US_TO_NS(((esc->dc_max - esc->dc_min) * speed) +
```

Comment:
Again, this is not duty cycle, but pulse duration, also the defines are not really needed and it would be much better to just create an intermediate variable like `pulse_duration_us` and then convert it accordingly later, either in the function call body or with another intermediate variable like `pulse_duration_ns`.